refactor: RPC: pass named argument value as string_view

stickies-v opened the pull request:
Inspired by MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26506#discussion_r1036149426). Main purpose of this PR is to minimize copying (potentially large) RPC named arguments when calling `.substr()` by using `std::string_view` instead of `std::string`. Furthermore, cleans up the code by removing unnecessary complexity in `ParseNonRFCJSONValue()` (done first to avoid refactoring required to concatenate `string` and `string_view`), updates some naming and adds a few test cases. Should not introduce any behaviour change.

## Questions
- ~Was there actually any merit to `ParseNonRFCJSONValue()` surrounding the value with brackets and then parsing it as an array? I don't see it, and the new approach doesn't fail any tests. Still a bit suspicious about it though.~
  - Cleared up by https://github.com/bitcoin/bitcoin/pull/26506#pullrequestreview-1211984059
  - If there are no objections to 7727603e44f8f674e0fc8389e78047e2b56e6052, I think we should follow up with a PR to rename `ParseNonRFCJSONValue()` to a local `Parse()` helper function (that throws if invalid), remove it from `client.h` and merge the test coverage we currently have on `ParseNonRFCJSONValue()` with the coverage we have on `UniValue::read()`.

stickies-v committed:
refactor: reduce unnecessary complexity in ParseNonRFCJSONValue

Since https://github.com/jgarzik/univalue/pull/31, UniValue::read() can now
parse raw literals directly, so there is no more need to wrap them into an
array first.

stickies-v committed:
refactor: use string_view for RPC named argument values

Minimize copying RPC named argument values when calling .substr() by
using std::string_view instead of std::string.

stickies-v committed:
test: add cases to JSON parsing

MarcoFalke commented on src/rpc/client.cpp:
```diff
@@ -225,11 +228,14 @@ class CRPCConvertTable
-        return (members.count(std::make_pair(method, idx)) > 0);
+    /* Return value as UniValue, and first parse it if it is a non-string parameter */
+    UniValue ParseIfNonString(std::string_view value, const std::string& method, int idx) {
+        return (members.count(std::make_pair(method, idx))) > 0 ? ParseNonRFCJSONValue(value) : value;
```
```suggestion
        return (members.count({method, idx})) > 0 ? ParseNonRFCJSONValue(value) : value;
```

nit

stickies-v commented:
Rebased now that #26506 is merged, ready for review!

Also added to PR description:

> If there are no objections to https://github.com/bitcoin/bitcoin/commit/7727603e44f8f674e0fc8389e78047e2b56e6052, I think we should follow up with a PR to rename ParseNonRFCJSONValue() to a local Parse() function, remove it from client.h and merge the test coverage we currently have on ParseNonRFCJSONValue() with the coverage we have on UniValue::read().

ryanofsky reviewed (approved):
Code review ACK 545ff924ab6303ffabd91fdfc4f0a4962daf133c

I think the followup ideas in the PR description sound very good too

pablomartin4btc reviewed (commented):
Approach ACK.

I've investigated why (if?) we would need to declare `const std::string_view`.
I see that while std::string_view is already read-only by design, declaring it as const can provide additional benefits in terms of communication, const correctness, and optimization but if we expect large values as @MarcoFalke [mentioned on his comment](https://github.com/bitcoin/bitcoin/pull/26506#discussion_r1035872217) perhaps we need to consider it.
This is a very interesting article on [reasons to pass std::string_view by value](https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/), there are links to [godbolt](https://godbolt.org/z/dvYnW6abK) snippets where it compares machine code differences on using one or the other approach.

Is there any benchmark that you think I could run in order to get some stats in terms of gains in performance improvement?

maflcko reviewed (approved):
nice

review ACK 545ff924ab6303ffabd91fdfc4f0a4962daf133c 📻

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 545ff924ab6303ffabd91fdfc4f0a4962daf133c 📻
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhxDQwAnkUGV9vWw6I02mle/ZFfeIWLL4UhIJjNmcc0UH+0DwtK7IM9V74ST0QZ
8iv5AOTv43xJ9Ltye54nIbJnjIhW1WmZ/nhwTHDqJtQHP2GqF4ao2uQo2nGMEt/G
O3BixrcDyYHQDOR92bou9pzc7BOJpQ3Imhbq12FmQrVFSDsREF9wyMU0sPm7OypZ
ZIBSZc8/7n3KzGFd4qwBwQHKP0Do4vngE/PnuW+OF5u8k8+xe1kny4KBSadccpNl
eeO/VTeH2hip8pgBweKG7iuWRvHNtsEnrRdHIKLpTQh13l1L4VaRiz/GzkeGDEGf
YdRDOI9f5CT3slSC93hEKODOakmWVxQ/7dvprogcefNpZSq1jYIrK1uK/z1ZSJ1u
FuvPBv6WZ4AhJkdFA9jzLKEZ/Fr6Bur0qeXJNQdoa7CpHK3tsEJ5KqhhFbjhrxlF
vDlABZ6iVD2S4hKF7keu7et+cxMSxsB6VvTd/3tI46TxZqGlyv7bE2VuPBGyiNst
pioPNtzT
=pTHK
-----END PGP SIGNATURE-----
```


</details>

fanquake commented:
> I think the followup ideas in the PR description sound very good too

@stickies-v will you follow up here?
